core/include/crypto/crypto_impl.h

```diff
@@ -297,4 +298,82 @@ drvcrypt_mac_alloc_ctx(struct crypto_mac_ctx **ctx __unused,
 	return TEE_ERROR_NOT_IMPLEMENTED;
 }
 #endif /* CFG_CRYPTO_DRV_MAC */
+
+/*
+ * The ECC public key operations used by the crypto_acipher_ecc_*() and
+ * crypto_acipher_free_ecc_*() functions.
+ * Reference set in ecc_public_key when key allocated.
+ */
+struct crypto_ecc_public_ops {
+	void (*free)(struct ecc_public_key *key);
+	TEE_Result (*verify)(uint32_t algo, struct ecc_public_key *key,
+			     const uint8_t *msg, size_t msg_len,
+			     const uint8_t *sig, size_t sig_len);
+};
+
+/*
+ * The ECC keypair operations used by the crypto_acipher_ecc_*() and
+ * crypto_acipher_gen_ecc_*() functions.
+ * Reference set in ecc_keypaipairr when key allocated.
+ */
+struct crypto_ecc_keypair_ops {
+	TEE_Result (*generate)(struct ecc_keypair *key, size_t key_size_bits);
+	TEE_Result (*sign)(uint32_t algo, struct ecc_keypair *key,
+			   const uint8_t *msg, size_t msg_len, uint8_t *sig,
+			   size_t *sig_len);
+	TEE_Result (*shared_secret)(struct ecc_keypair *private_key,
+				    struct ecc_public_key *public_key,
+				    void *secret, unsigned long *secret_len);
+};
+
+#define CRYPTO_ASYM_KEY_ALLOC_NOT_IMPLEMENTED(name) \
+	static inline TEE_Result crypto_asym_alloc_##name##_public_key( \
+		struct name##_public_key *key __unused, \
+		uint32_t key_type __unused, size_t key_size_bits __unused) \
+	{ \
+		return TEE_ERROR_NOT_IMPLEMENTED; \
+	} \
+	static inline TEE_Result crypto_asym_alloc_##name##_keypair( \
+		struct name##_keypair *key __unused, \
+		uint32_t key_type __unused, size_t key_size_bits __unused) \
+	{ \
+		return TEE_ERROR_NOT_IMPLEMENTED; \
+	}
+
+#define DRVCRYPT_ASYM_KEY_ALLOC_NOT_IMPLEMENTED(name) \
+	static inline TEE_Result drvcrypt_asym_alloc_##name##_public_key( \
+		struct name##_public_key *key __unused, \
+		uint32_t key_type __unused, size_t key_size_bits __unused) \
+	{ \
+		return TEE_ERROR_NOT_IMPLEMENTED; \
+	} \
+	static inline TEE_Result drvcrypt_asym_alloc_##name##_keypair( \
+		struct name##_keypair *key __unused, \
+		uint32_t key_type __unused, size_t key_size_bits __unused) \
+	{ \
+		return TEE_ERROR_NOT_IMPLEMENTED; \
+	}
```

Comment:
I don't like these macros much for 3 reasons.
- First, they obfuscate the code and make it harder to parse (editor go to definition/declaration, grep etc.) for very little benefits (a few lines of code saved).
- Second, they are very similar (`crypto_` vs. `drvcrypt`), so they only go half way in the direction of saving lines.
- Third, by the same logic you could use macros to declare the prototypes too (the `ifdef` parts).

Reply:
I defined those macro because of the other asymmetric algorithm (DSA, RSA, DH, ..) that will have to be modified too, using the same logic.

I copycat what has been done the SW library function declaration as done for algo like hash, cipher). And in same time I did the same type of macro for the Crypto HW driver.

```
ifdef CFG_CRYPTO_XXX
TEE_Result crypt_asym_alloc_XXX_public_key(struct XXX_public_key *key,
					      uint32_t key_type,
					      size_t key_size_bits);
TEE_Result crypt_asym_alloc_XXX_keypair(struct XXX_keypair *key,
					   uint32_t key_type,
					   size_t key_size_bits);
else
static inline TEE_Result crypt_asym_alloc_XXX_public_key(struct XXX_public_key *key,
					      uint32_t key_type,
					      size_t key_size_bits) 
{
     return TEE_ERROR_NOT_IMPLEMENTED
};
static inline TEE_Result crypt_asym_alloc_XXX_keypair(struct XXX_keypair *key,
					   uint32_t key_type,
					   size_t key_size_bits)
{
     return TEE_ERROR_NOT_IMPLEMENTED
};
endif
```

But if you prefer I can keep the crypto HW driver detailed like done for hash, cipher.

Reply:
I prefer the detailed way too.